File: Practical4.py
'''
Practical:- 4
Aim: Implement a solution for Constraint Satisfaction Problem
using Backtracking and Branch & Bound for N-Queens Problem

Python Program for N-Queens Problem using Branch and Bound

'''
class NQueens:
    def __init__(self, n):
        self.n = n

        # Create chessboard initialized with 0
        self.board = [[0 for _ in range(n)] for _ in range(n)]

        # Arrays for Branch and Bound optimization
        self.rows = [False] * n
        self.upperDiagonal = [False] * (2 * n - 1)
        self.lowerDiagonal = [False] * (2 * n - 1)

    # Function to solve N-Queens problem
    def solveNQUtil(self, col):

        # Base Case: If all queens are placed
        if col >= self.n:
            return True

        # Try placing queen in all rows one by one
        for i in range(self.n):

            # Check if queen can be placed safely
            if (self.rows[i] == False and
                self.lowerDiagonal[i + col] == False and
                self.upperDiagonal[self.n - 1 + col - i] == False):

                # Place the queen
                self.board[i][col] = 1

                # Mark row and diagonals as occupied
                self.rows[i] = True
                self.lowerDiagonal[i + col] = True
                self.upperDiagonal[self.n - 1 + col - i] = True

                # Recur to place rest of the queens
                if self.solveNQUtil(col + 1):
                    return True

                # BACKTRACKING
                # Remove queen if solution not found
                self.board[i][col] = 0

                # Unmark row and diagonals
                self.rows[i] = False
                self.lowerDiagonal[i + col] = False
                self.upperDiagonal[self.n - 1 + col - i] = False

        # If queen cannot be placed in any row
        return False

    # Function to print solution
    def printSolution(self):

        if self.solveNQUtil(0) == False:
            print("Solution does not exist")
            return

        print(f"\nSolution for {self.n}-Queens Problem:\n")

        for i in range(self.n):
            for j in range(self.n):
                print(self.board[i][j], end=" ")
            print()

        print("\nChessboard Representation:\n")

        for i in range(self.n):
            for j in range(self.n):
                if self.board[i][j] == 1:
                    print("Q", end=" ")
                else:
                    print(".", end=" ")
            print()


# Driver Code
if __name__ == "__main__":

    n = int(input("Enter value of N: "))

    queens = NQueens(n)
    queens.printSolution()

'''

Enter value of N: 4

Solution for 4-Queens Problem:

0 0 1 0
1 0 0 0
0 0 0 1
0 1 0 0

Chessboard Representation:

. . Q .
Q . . .
. . . Q
. Q . .


'''

File: Practical4.cpp
/*
Practical 4:
Aim: Implement a solution for Constraint Satisfaction Problem
using Backtracking and Branch & Bound for N-Queens Problem.
*/
/*
Sample Input:

Enter value of N: 4

Sample Output:

Solution for 4-Queens Problem:

0 0 1 0
1 0 0 0
0 0 0 1
0 1 0 0

Chessboard Representation:

. . Q .
Q . . .
. . . Q
. Q . .
*/

#include <iostream>
#include <vector>
using namespace std;

class NQueens
{
private:
    int n;
    vector<vector<int>> board;

    // Arrays for Branch and Bound optimization
    vector<bool> rows;
    vector<bool> upperDiagonal;
    vector<bool> lowerDiagonal;

public:
    // Constructor
    NQueens(int size)
    {
        n = size;

        board = vector<vector<int>>(n, vector<int>(n, 0));

        rows = vector<bool>(n, false);
        upperDiagonal = vector<bool>(2 * n - 1, false);
        lowerDiagonal = vector<bool>(2 * n - 1, false);
    }

    // Function to solve N-Queens problem
    bool solveNQUtil(int col)
    {
        // Base Case: If all queens are placed
        if (col >= n)
        {
            return true;
        }

        // Try placing queen in all rows one by one
        for (int i = 0; i < n; i++)
        {
            /*
            Check if queen can be placed safely:
            1. Row should not be occupied
            2. Lower diagonal should not be occupied
            3. Upper diagonal should not be occupied
            */

            if (rows[i] == false &&
                lowerDiagonal[i + col] == false &&
                upperDiagonal[n - 1 + col - i] == false)
            {
                // Place the queen
                board[i][col] = 1;

                // Mark row and diagonals as occupied
                rows[i] = true;
                lowerDiagonal[i + col] = true;
                upperDiagonal[n - 1 + col - i] = true;

                // Recur to place rest of the queens
                if (solveNQUtil(col + 1))
                {
                    return true;
                }

                // BACKTRACKING
                // Remove queen if solution not found
                board[i][col] = 0;

                // Unmark row and diagonals
                rows[i] = false;
                lowerDiagonal[i + col] = false;
                upperDiagonal[n - 1 + col - i] = false;
            }
        }

        // If queen cannot be placed in any row
        return false;
    }

    // Function to print solution
    void printSolution()
    {
        if (solveNQUtil(0) == false)
        {
            cout << "Solution does not exist" << endl;
            return;
        }

        cout << "\nSolution for " << n << "-Queens Problem:\n\n";

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                cout << board[i][j] << " ";
            }
            cout << endl;
        }

        cout << "\nChessboard Representation:\n\n";

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (board[i][j] == 1)
                {
                    cout << "Q ";
                }
                else
                {
                    cout << ". ";
                }
            }
            cout << endl;
        }
    }
};

// Driver Code
int main()
{
    int n;

    cout << "Enter value of N: ";
    cin >> n;

    NQueens queens(n);
    queens.printSolution();

    return 0;
}